multi/gr716-multi/adc.c

```diff
@@ -0,0 +1,261 @@
+	adc_seqSync,    /* Sequence sync register : 0x0C */
+	adc_status,     /* Status register : 0x10 */
+	adc_irq,        /* Interrupt register : 0x14 */
+	adc_irqMsk,     /* Interrupt mask register : 0x18 */
+	adc_preAmp,     /* Pre-amplifier control register : 0x1C */
+	adc_hiLvl,      /* High level detection register : 0x20 */
+	adc_loLvl,      /* Low level detection register : 0x24 */
+	adc_seqSts = 11 /* Sequence status register : 0x2C - 0x38 */
+};
+
+
+static const struct {
+	volatile uint32_t *base;
+	uint32_t irq;
+	int active;
+} adc_info[ADC_CNT] = {
+	{ ADC0_BASE, ADC0_IRQ, 1 },
+	{ ADC1_BASE, ADC1_IRQ, 0 },
+	{ ADC2_BASE, ADC2_IRQ, 0 },
+	{ ADC3_BASE, ADC3_IRQ, 0 },
+	{ ADC4_BASE, ADC4_IRQ, 0 },
+	{ ADC5_BASE, ADC5_IRQ, 0 },
+	{ ADC6_BASE, ADC6_IRQ, 0 },
+	{ ADC7_BASE, ADC7_IRQ, 0 }
+};
+
+
+static struct {
+	handle_t mutex;
+	handle_t irqLock;
+	handle_t cond;
+	handle_t inth;
+} adc_common[ADC_CNT];
+
+
+static int adc_irqHandler(unsigned int n, void *arg)
+{
+	int adc = (int)arg;
+	volatile uint32_t *adc_base = adc_info[adc].base;
+
+	(void)n;
+
+	if ((*(adc_base + adc_irq) & ADC_IRQ_CONV_END) != 0) {
+		*(adc_base + adc_irq) = ADC_IRQ_CONV_END;
+		return 1;
+	}
+
+	return 0;
+}
+
+
+static void adc_convert(int dev, uint32_t *value)
+{
+	volatile uint32_t *adc_base = adc_info[dev].base;
+
+	mutexLock(adc_common[dev].mutex);
+
+	/* Start conversion */
+	common_atomicOr(ADC_CONV_START, adc_base + adc_ctrl);
+
+	/* Wait for conversion to end */
+	while ((*(adc_base + adc_status) & ADC_STS_CONV_END) == 0) {
+		condWait(adc_common[dev].cond, adc_common[dev].mutex, 0);
+	}
+
+	*value = *(adc_base + adc_status) & ADC_STS_VAL;
+
+	mutexUnlock(adc_common[dev].mutex);
+}
+
+
+static void adc_configure(int dev, adc_config_t *config)
+{
+	volatile uint32_t *adc_base = adc_info[dev].base;
+	uint32_t scaler = (SYSCLK_FREQ / (config->sampleRate * 20));
+
+	uint32_t ctrl = (scaler << 16) | ((dev & 0xf) << 2) | (config->mode & 0x1) | ADC_ENABLE;
+
+	mutexLock(adc_common[dev].mutex);
+
+	*(adc_base + adc_ctrl) = ctrl;
+
+	/* Disable pre-amp */
+	*(adc_base + adc_preAmp) = ADC_PREAMP_BYPASS;
+
+	/* Enable end of conversion interrupt */
+	*(adc_base + adc_irqMsk) = ADC_IRQ_CONV_END;
+
+	mutexUnlock(adc_common[dev].mutex);
+}
+
+
+static void adc_handleDevCtl(msg_t *msg, int dev)
+{
+	multi_i_t *idevctl = (multi_i_t *)msg->i.raw;
+	multi_o_t *odevctl = (multi_o_t *)msg->o.raw;
+
+	dev -= id_adc0;
+	if ((dev >= ADC_CNT) || adc_info[dev].active == 0) {
+		odevctl->err = -EINVAL;
+		return;
+	}
+
+	switch (idevctl->adc.type) {
+		case adc_conversion:
+			adc_convert(dev, &odevctl->val);
+			odevctl->err = EOK;
+			break;
+
+		case adc_config:
+			adc_configure(dev, &idevctl->adc.config);
+			odevctl->err = EOK;
+			break;
+
+		default:
+			odevctl->err = -EINVAL;
+			break;
+	}
+}
+
+
+void adc_handleMsg(msg_t *msg, int dev)
+{
+	switch (msg->type) {
+		case mtOpen:
+		case mtClose:
+		case mtWrite:
+		case mtRead:
```

Comment:
We could handle `mtRead` based on `msg->i.io.oid.id`. Ofc we need a file in /dev per each adc